### ISSUE TITLE
tf.data performance slow when returning many tensors from a dataset op

### ISSUE THREAD

**Issue by akyrola** [tensorflow/tensorflow#25263]
**System information**
- Linux Ubuntu 16.04
- i7-7800X CPU @ 3.50GHz x 8
- Tensorflow binary, r1.12
- TensorFlow version (use command below): ('v1.12.0-0-ga6d8ffae09', '1.12.0')
- Python version: Python 2.7.12

**Describe the current behavior**

We load tabular data from a database to a training script, and this tabular data contains quite a many columns (30 - 40), most of them scalar. We observe performance that seems slow, given the simplicity of the data. To demonstrate and isolate the performance bottleneck, I created a simple C++ dataset op that returns just a specified number of int scalar tensors, and we get only about 1000 - 1500 entries / second  (calls of get_next) with this fake data.  These overheads lead to high CPU utilization and limit the throughput of our training.

My code with more results can be seen here:
https://github.com/akyrola/tf_data_repro

The key code is simply:
```
        for (int i = 0; i < dataset()->n_; ++i) {
                    tensorflow::Tensor tensor(ctx->allocator({}), tensorflow::DT_INT32, shape);
                    tensor.scalar<tensorflow::int32>()() = row;
                    out_tensors->emplace_back(std::move(tensor));
                }
```

The C++ op is compiled with -O2 optimizations.

I am suspecting the overhead is due to the tensor allocations, because the throughput changes almost linearly with the number of tensors (when n sufficiently large). I am looking for advice if using different allocator would help, for example.

**Describe the expected behavior**

I would expect at least 10x better performance on fake data generation.

**Code to reproduce the issue**

See code checked here for a benchmark:
https://github.com/akyrola/tf_data_repro

**Other info / logs**
Benchmark results (approximate), where n = number of scalar tensors:
- n=1:   8300 rows/sec
- n=2:   6800 rows/sec
- n=5:   4800 rows/sec
- n=10:  3200 rows/sec
- n=20:  2000 rows/sec
- n=40:  1100 rows/sec

**Comment by jsimsa** [comment]
Hey Aapo,

TLDR: You are not benchmarking TensorFlow's ability to allocate tensors, your are benchmarking TensorFlow's ability to execute a graph with n outputs.

If you would like to measure the performance of the inner loop of your dummy dataset, try the following:

```
def benchmark(n):
    num_rows = 1000 * 1000
    dataset = dummy_dataset.DummyDataset(n).skip(num_rows - 1)
    iterator = dataset.make_one_shot_iterator()
    get_next = iterator.get_next()

    with tf.Session() as sess:
      sess.run(get_next) # warm up caches
      start_time = time.time()
      row = sess.run(get_next)

    print(row)
    print("Finished (n={}): read {} rows, {} rows / sec".format(
        n, num_rows, num_rows / (time.time() - start_time)))
```

Using the `skip` transformation will result in running the `GetNext` call of your dummy dataset in a tight loop `num_rows` number of times.

When I do this on my workstation, I see throughput upwards of 100M rows/sec (compared to 1K rows/sec reported by your benchmarking code).

**Comment by akyrola** [comment]
Hi Jiri,

thanks for your reply, this was indeed very clear way to isolate the overhead to the tensorflow graph execution. I guess we need to reduce the number of tensors we use and also limit the depth of the tf.data pipeline as each processing step adds a lot of overhead due to the number of tensors involved.

**Comment by jsimsa** [comment]
Hi Aapo, note that you do not need to reduce the tf.data input pipeline depth as it does not contribute to the graph size -- fetching of an element from an arbitrary tf.data input pipeline corresponds to executing a single graph node.

Perhaps you could batch the data?
